codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java

```diff
@@ -265,18 +271,21 @@ public Http2Stream resetSent() {
         }
 
         @Override
-        public final Object setProperty(Object key, Object value) {
-            return data.put(key, value);
+        public final <K, V> V setProperty(PropertyKey<K, V> key, V value) {
+            checkNotNull(key, "key");
+            return properties.add((DefaultPropertyKey<K, V>) key, value);
```

Comment:
So I think this can reasonably be pulled off with the following:
1. Update the PropertyKey interface as described in https://github.com/netty/netty/pull/3593#discussion_r28541081
2. Define an enum (or something) in the codec for uses of the property map that can also serve as the `index()` for that purpose.  The `PropertyKeyRegistry` by default should start assigning new properties one more than the highest index.
3. Have `PropertyKeyRegistry` extensible such that the "base" index can be configured so users can also reserve indices.